src/fs_watcher.rs

```diff
@@ -0,0 +1,124 @@
+        std::fmt::Display::fmt(&self.reason, f)
+    }
+}
+
+impl std::error::Error for FsWatcherError {}
+
+impl FsWatcher {
+    pub fn new(endpoints_dir: PathBuf, dst_sender: DstSender) -> Self {
+        Self {
+            endpoints_dir,
+            dst_sender,
+        }
+    }
+
+    fn parse_dst(file_name: &str) -> Result<(Dst, FileType), Error> {
+        let mut parts = file_name.rsplitn(2, ".");
+        match (parts.next(), parts.next()) {
+            (Some(ext), Some(name)) => {
+                let dst: Dst = name.parse()?;
+                let ft = match ext {
+                    "yaml" => FileType::Yaml,
+                    "json" => FileType::Json,
+                    _ => fs_watcher_error!(format!("invalid file ext {}", ext)),
+                };
+                Ok((dst, ft))
+            }
+            _ => fs_watcher_error!(format!("invalid file name {}", file_name)),
+        }
+    }
+
+    fn parse_file(&self, file_name: &str, ft: FileType) -> Result<Endpoints, Error> {
+        let path = self.endpoints_dir.join(file_name);
+        let contents = fs::read_to_string(path)?;
+        let destinations = match ft {
+            FileType::Json => {
+                serde_json::from_str::<Vec<EndpointMeta>>(&contents).map_err(Into::into)
+            }
+            FileType::Yaml => {
+                serde_yaml::from_str::<Vec<EndpointMeta>>(&contents).map_err(Into::into)
+            }
+        };
+
+        destinations.map(|dsts| Endpoints(dsts.into_iter().map(|e| (e.address, e)).collect()))
+    }
+
+    #[tracing::instrument(skip(self), name = "FsWatcher::handle_event", level = "info")]
+    async fn handle_event(&mut self, ev: Event<OsString>) -> Result<(), Error> {
+        if let Some(file_name) = ev.name.and_then(|s| s.to_str().map(|s| s.to_string())) {
+            let (dst, ft) = Self::parse_dst(&file_name)?;
+            if ev.mask == EventMask::DELETE {
+                tracing::info!(?dst, "deleted");
+                self.dst_sender.delete_dst(dst).await;
+            } else {
+                let endpoints = self.parse_file(&file_name, ft)?;
+                tracing::info!(?endpoints, "added");
+                self.dst_sender.send_endpoints(dst, endpoints).await?;
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn watch(&mut self) -> Result<(), Error> {
+        let mut inotify = Inotify::init()?;
+        let mask = WatchMask::MODIFY | WatchMask::DELETE;
+        inotify.add_watch(self.endpoints_dir.clone(), mask)?;
+        let mut stream = inotify.event_stream(vec![0; EVENT_BUF_SZ])?;
+        while let Some(event) = stream.next().await {
+            match event {
+                Ok(event) => {
+                    if let Err(e) = self.handle_event(event).await {
+                        tracing::error!(?e, "error handing event");
+                    }
+                }
+                Err(e) => tracing::error!(?e, "inotify stream error"),
```

Comment:
nit: I'm pretty sure all these errors should implement `fmt::Display`, so we can format them using `Display` instead:
```suggestion
                        tracing::error!(%e, "error handing event");
                    }
                }
                Err(e) => tracing::error!(%e, "inotify stream error"),
```